### PR TITLE
OCD-2505-Verify Implementation Note update for certified_products end…

### PR DESCRIPTION
**Pull request by ashwinimore**
…point

### CODE REVIEW

File: src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public final class ChplAPIPage {
+    private ChplAPIPage() {}
+    private static WebElement element = null;
+
+    /**
+     * Returns element holding main content.
+     * @param driver WebDriver
+     * @return element holding main content
+     */
+    public static WebElement mainContent(final WebDriver driver) {
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java]: I wonder if we could do something where we have a "BasePage" class that has this element in it, along with the navigation ones, and have all of our specific pages extend from that. Then all pages would have a reference to the main-content as well as navigation things, and we'd have the code in only one place

File: src/test/java/gov/healthit/chpl/aqa/stepDefinitions/ChplAPIDocumentationSteps.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import cucumber.api.java.en.And;
+import cucumber.api.java.en.Given;
+import cucumber.api.java.en.Then;
+import cucumber.api.java.en.When;
+import gov.healthit.chpl.aqa.pageObjects.ChplAPIPage;
+
+/**
+ * Class ChplAPIDocumentationSteps definition.
+ */
+public class ChplAPIDocumentationSteps {
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/stepDefinitions/ChplAPIDocumentationSteps.java]: Take a look at "ChplDownloadSteps" and "ManageDevelopersAndProductsSteps" for examples of steps pages that extend the BaseSteps class. By extending BaseSteps you don't have to manage the webdriver, url, timeouts, etc.

File: src/test/resources/Features/APIDocumentationUpdate.feature [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Feature: API Documentation updates
+         OCD-2505 - Update Implementation Note for /certified_products endpoint
+
+Scenario Outline: Verify that implementation note for /certified_products endpoint is updated with new text
+    Given I am on CHPL API page
+    When I click on "<controller_name>" to see operations list
+    And I click on Certified Products endpoint link to view details
```
Comment (andlar) [inline src/test/resources/Features/APIDocumentationUpdate.feature]: Either the "When" and "Then" steps shouldn't be parameterized, or the "And" step should. There's no way to do a second Example without changing the "And".

The "solves the feature" option would be to not parameterize either of them, while the "prepares for the future" option would be to do it for them all. If you _do_ do it for them all, I'd recommend putting an Example in that should have been successful even before the story started

**Review by andlar — COMMENTED** [review]
I have some comments, but don't think they're required changes. If you want to address them that's great, if not consider this an approval

File: src/test/java/gov/healthit/chpl/aqa/stepDefinitions/BaseSteps.java [hunk 4]
```diff
@@ -63,4 +66,15 @@ public WebDriverWait getWait() {
     public void setWait(final WebDriverWait wait) {
         this.wait = wait;
     }
+
+    /**
+     * Returns element holding main content.
+     * @param driver WebDriver
+     * @return element holding main content
+     */
+    public static WebElement mainContent(final WebDriver driver) {
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/stepDefinitions/BaseSteps.java]: I like this, but it's in the wrong file. It should be in a new "BasePage.java" file in the pages directory, not in the steps file

**Review by andlar — CHANGES_REQUESTED** [review]
Love the changes. Just need to move the Page part into a BasePage class and we're set. Might add another Example to the Feature just for completeness, but that's not required.

File: src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package gov.healthit.chpl.aqa.pageObjects;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+/**
+ * Class ChplAPIPage definition.
+ */
+public final class ChplAPIPage {
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java]:
```suggestion
public final class ChplAPIPage extends BasePage {
```

This class needs to extend the BasePage, then any calling step can do `ChplApiPage.mainContent` and get that element

File: src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private ChplAPIPage() {}
+    private static WebElement element = null;
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java]:
```suggestion
```
This isn't needed

File: src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+     /**
+     * Controller link.
+     * @param driver WebDriver
+     * @return controller link element
+     * @param controller desired controller to open
+     */
+    public static WebElement controllerLink(final WebDriver driver, final String controller) {
+        element = driver.findElement(By.xpath("//*[@id=\"" + controller + "*\"]/li/h4/a"));
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/pageObjects/ChplAPIPage.java]:
```suggestion
        return driver.findElement(By.xpath("//*[@id=\"" + controller + "*\"]/li/h4/a"));
```
Can delete the next line, and do the same changes for the other PageObject elements

File: src/test/java/gov/healthit/chpl/aqa/stepDefinitions/ChplAPIDocumentationSteps.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+       }
+
+    /**
+     * Get user to the API Documentation page.
+     */
+    @Given("^I am on CHPL API page$")
+    public void userLoadsAPIPage() {
+        getDriver().get(getUrl() + "#/resources/chpl_api");
+        WebDriverWait wait = new WebDriverWait(getDriver(), TIMEOUT);
+        wait.until(ExpectedConditions.visibilityOf(BasePage.mainContent(getDriver())));
```
Comment (andlar) [inline src/test/java/gov/healthit/chpl/aqa/stepDefinitions/ChplAPIDocumentationSteps.java]:
```suggestion
        wait.until(ExpectedConditions.visibilityOf(ChplAPIPage.mainContent(getDriver())));
```
With ChplApiPage extending BasePage, you can treat the ChplApiPage as having all of the BasePage content

**Review by andlar — CHANGES_REQUESTED** [review]

**Review by andlar — APPROVED** [review]

**Review by kekey1 — APPROVED** [review]

**Pull request merged**